### PR TITLE
Ensure all packages declare `package-info.java` with null-safety annotations

### LINKED ISSUE #30056: Ensure all packages declare `package-info.java` with null-safety annotations
## Overview

Since Spring Framework 5.0 (see #20099), we annotate packages in `package-info.java` files with null-safety annotations such as `@NonNullApi`; however, not all packages contain `package-info.java` files which prevents us from enforcing non-null semantics by default for those packages.

## Scope

This applies to all packages within `src/main`. Conversely, this does not apply to packages in `src/test`.

## Deliverables

- [ ] Configure the [`JavadocPackage`](https://checkstyle.sourceforge.io/config_javadoc.html#JavadocPackage) Checkstyle module to require `package-info.java` files for all packages under `src/main`.
- [ ] Determine if it is possible to require that `package-info.java` files include null-safety annotations – for example, via Checkstyle.
- [ ] If it's possible to require that `package-info.java` files include null-safety annotations, configure the necessary infrastructure.
- [ ] Introduce missing `package-info.java` files with package-level Javadoc and null-safety annotations.

### PR DESCRIPTION
**Pull request by edyda99**
## Overview

This commit adds the `JavadocPackage` Checkstyle module to the project's configuration file (`checkstyle.xml`). The module is configured to emit a warning message when a `package-info.java` file is missing, rather than throwing an exception. In addition, suppressions have been added to prevent the `JavadocPackage` checks from being applied to packages outside of the `src/main` directory. These suppressions ensure that the module only checks the packages that are relevant to the project and prevents false positives from being reported.

## Deliverables

- [x] Configure the [`JavadocPackage`](https://checkstyle.sourceforge.io/config_javadoc.html#JavadocPackage) Checkstyle module to require `package-info.java` files for all packages under `src/main`.
- [x] Determine if it is possible to require that `package-info.java` files include null-safety annotations – for example, via Checkstyle.
- [x] If it's possible to require that `package-info.java` files include null-safety annotations, configure the necessary infrastructure.
- [x] Introduce missing `package-info.java` files with package-level Javadoc and null-safety annotations.

### CODE REVIEW

**Review by simonbasle — COMMENTED** [review]
thanks for the suggestion. Can you provide more details about the choices you made for checkstyle modules?

**Comment by simonbasle** [comment]
(didn't see the issue reference, which probably provides some context. my bad)

**Comment by simonbasle** [comment]
@edyda99 I think the `RegexpJavaSingleLine` variant does the job (unless you've spotted occurrences of multilines nullability annotations?) The other one is also harder to read and has its own caveats with comments, so I'd go with the first one.

**Comment by edyda99** [comment]
Hi @simonbasle, I will go with RegexpJavaSingleLine.
No, I didn't spot any occurrences of multilines nullability, but it was about preventing it from happening.
So should I remove the whitespace warning? ( I would suggest keeping it and increasing the severity to "**error**", so whenever someone writes `@\n<annotation_name>` an error will be thrown)

**Comment by simonbasle** [comment]
> ( I would suggest keeping it and increasing the severity to "**error**", so whenever someone writes `@\n<annotation_name>` an error will be thrown)

Perhaps a simpler one line regexp like `@\s*$` (at then any number of whitespace then the end of the line) would fit the bill? That said, I think this PR can focus on the actual null-safety annotations.

**Comment by simonbasle** [comment]
>     * Do you suggest doing micro commits for each missing package-info.java?

Not necessarily, if that's the only thing in the commit.

>     * What do I need to rely on before adding null-safety annotations on a newly added package-info?

It depends on how many package-info are missing. I would locally prepare the change and run some static analysis trying to find null-safety issues raising from that change. If the build doesn't break, I'm not sure I would make it a goal of this PR to fix the null-safety warnings though. Wdyt @sbrannen (also cc @sdeleuze)

**Comment by sbrannen** [comment]
> > * Do you suggest doing micro commits for each missing package-info.java?
> 
> Not necessarily, if that's the only thing in the commit.

I agree with @simonbasle.

Ideally the final result will be 3-4 commits.

- enforce that `package-info.java` files exist and contain Javadoc
  - build should fail
- introduce missing `package-info.java` files
  - build should pass
- enforce that `package-info.java` files declare `@NonNullApi` and `@NonNullFields`
  - build should fail
- introduce missing `@NonNullApi` and `@NonNullFields` declarations
  - build should pass

Of course, we may squash all commits in the PR into a single commit if that turns out to be easier.

> > * What do I need to rely on before adding null-safety annotations on a newly added package-info?

I don't understand the question. Can you please expound?

> It depends on how many package-info are missing. I would locally prepare the change and run some static analysis trying to find null-safety issues raising from that change. If the build doesn't break, I'm not sure I would make it a goal of this PR to fix the null-safety warnings though. Wdyt @sbrannen (also cc @sdeleuze)

Ensuring that proper null-safety semantics are enforced _within_ code in newly annotated packages is beyond the scope of this PR/issue. I'll create a follow-up issue to address that.

File: src/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -239,5 +239,21 @@
...
- 	</module>
+		<!--package-info should contain null-safety annotations-->
+		<!--This module will fail to detect multiline annotations-->
+		<module name="RegexpSinglelineJavaCheck">
+			<property name="id" value="javaDocPackageAnnotations"/>
+			<property name="format" value="(@NonNull|@org\.springframework\.lang\.NonNull)"/>
+			<property name="severity" value="warning"/>
+			<property name="minimum" value="1"/>
+			<property  name="maximum" value="5"/>
+			<property name="message" value="package-info.java is missing null-safety annotations  or has null-safe annotations made up of multiple lines. Please check."/>
```
Comment (sbrannen) [inline src/checkstyle/checkstyle.xml:250]:
```suggestion
			<property name="message" value="package-info.java is missing required null-safety annotations @NonNullApi and @NonNullFields."/>
```

Spring's `@NonNullApi` and `@NonNullFields` annotations should never span multiple lines.

**Review by sbrannen — CHANGES_REQUESTED** [review]
Thanks for working on this!

It's coming along nicely.

I've requested a few minor changes.

After that, all that's left is ensuring that the build fails for violations and that all missing `package-info.java` files have been created (so that the build passes again).

File: src/checkstyle/checkstyle.xml [hunk 2]
```diff
@@ -239,5 +239,21 @@
...
+			<property name="ignoreComments" value="true"/>
+		</module>
+	</module>
+
+	<!--package-info checker -->
+	<module name="JavadocPackage">
+		<property name="severity" value="warning"/>
```
Comment (sbrannen) [inline src/checkstyle/checkstyle.xml:257]: We also want these violations to fail the build.

**Comment by sbrannen** [comment]
You also need to suppress violations in the `framework-docs` module, since that source code is only used in the reference manual.

**Comment by edyda99** [comment]
Hi again,

First sorry for the continuous Force pushes, but I wanted to reduce my commits.
I reduced them to 4 micro commits, with the last one containing the full commit message.
All comments are resolved.

The merge request is ready for the final review

Thank you again!

**Comment by edyda99** [comment]
@sbrannen Ok Great
Could you suggest something else for me?

**Comment by sbrannen** [comment]
This has been merged into `main` in 268e3fec9921653de17ab2f495039239f6b6152e and 7e32f504b021015ca8a7eaed7280453fad724b0f and revised and completed in 99e54fec3ac8dfedd2537fde5f19fbcc4f1f6202.

The latter turned out to be a bit more work than expected. 👀 

Thanks again for the contribution, @edyda99!

**Pull request closed without merging**